compaction.go

```diff
@@ -85,30 +64,38 @@ func (f *FileCompaction) writeRecordsToParquetFile(compact []parts.Part, options
 		return nil, 0, 0, fmt.Errorf("no parts to compact")
 	}
 
-	accountant := &accountingWriter{w: f.file}
-	preCompactionSize, err := f.t.compactParts(accountant, compact,
-		parquet.KeyValueMetadata(
-			ParquetCompactionTXKey, // Compacting up through this transaction.
-			fmt.Sprintf("%v", compact[0].TX()),
-		),
-	) // compact into the next level
-	if err != nil {
+	tx := compact[0].TX()
+	fileName := filepath.Join(f.dir, fmt.Sprintf("%v.parquet", tx))
+	var preCompactionSize int64
+	accountant := &accountingWriter{}
+	if err := func() error {
+		file, err := os.OpenFile(fileName, os.O_CREATE|os.O_RDWR, filePerms)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		accountant.w = file
+		preCompactionSize, err = f.t.compactParts(accountant, compact,
+			parquet.KeyValueMetadata(
+				ParquetCompactionTXKey, // Compacting up through this transaction.
+				fmt.Sprintf("%v", compact[0].TX()),
+			),
+		) // compact into the next level
+		if err != nil {
+			return err
+		}
+		return nil
+	}(); err != nil {
 		return nil, 0, 0, err
 	}
 
-	// Record the writing offset into the file.
-	prevOffset := f.offset
-	f.offset += accountant.n + 8
-	f.ref++
-
-	// Record the file size for recovery.
-	size := make([]byte, 8)
-	binary.LittleEndian.PutUint64(size, uint64(accountant.n))
-	if n, err := f.file.Write(size); n != 8 {
-		return nil, 0, 0, fmt.Errorf("failed to write size to file: %v", err)
+	file, err := os.Open(filepath.Join(f.dir, fmt.Sprintf("%v.parquet", tx)))
```

Comment:
I don't know if this is going to be tenable to maintain open files. I think we'll hit the open file descriptor limit.

Instead I think we need to change the Part to just maintain the filename and only open when it's read.

Reply:
But also trying to track the lifecycle during a read is going to be challenging

Reply:
So maybe we just leave it as open fd and then control for open fd by adjusting level sizing. wdyt @asubiotto?

Reply:
I think that LSMs like RocksDB and pebble keep file descriptors around up to a pre-configured limit. IIUC, these file descriptors are kept in a table cache (see https://github.com/facebook/rocksdb/wiki/RocksDB-Overview#table-cache and https://github.com/cockroachdb/pebble/blob/8ce58a8116e95f2db3c2c77f797c1ea40d57d536/table_cache.go#L271).

I think it might be simpler to start by having a well defined descriptor lifecycle. Can we somehow piggy back off of the parquet `Rows` interface to do this? I would have to take a closer look.

Reply:
Additionally I think we should also take a closer look at level sizing now since as you say this affects the number of files. Specifically, taking RocksDB as inspiration, I think there's a setting called `target_file_size_base` which defaults to 64MB and specifies the size of a single SST file (can also configure this size per level with a `target_file_size_base_multiplier` or something). We could add this option and change compactions to merge overlapping files to limit the number of files per level.

Reply:
Yea I think we can wrap row groups with a reference counter to open and close files accordingly to control the lifecycle of fd.

Reply:
I think I'll leave managing the number of files to a future PR should we feel the need.